alica_engine/src/engine/RunnableObject.cpp

```diff
@@ -34,70 +38,125 @@ ThreadSafePlanInterface RunnableObject::getPlanContext() const
     return ThreadSafePlanInterface(isExecutingInContext() ? _execContext.load() : nullptr);
 }
 
-void RunnableObject::stop()
+void RunnableObject::doStop()
 {
     if (!isActive(_signalState.load())) {
         return;
     }
     ++_signalState;
-    _engine->editScheduler().schedule(std::bind(&RunnableObject::doTerminate, this));
 }
 
-void RunnableObject::start(RunningPlan* rp)
+void RunnableObject::doStart(RunningPlan* rp)
 {
     if (isActive(_signalState.load())) {
         return;
     }
     ++_signalState;
     _signalContext.store(rp);
-    _engine->editScheduler().schedule(std::bind(&RunnableObject::doInit, this));
 }
 
-void RunnableObject::setTerminatedState()
+void RunnableObject::startTrace()
 {
-    ++_execState;
-
-    clearFlags(Flags::RUN_TRACED);
-    if (!areFlagsSet(Flags::INIT_EXECUTED)) {
-        _execContext.store(nullptr);
+    if (!_engine->getTraceFactory()) {
         return;
     }
-    clearFlags(Flags::INIT_EXECUTED);
-}
 
-void RunnableObject::traceTermination()
-{
-    if (_trace) {
-        _trace->setLog({"Terminate", "true"});
-        _trace.reset();
+    switch (_tracingType) {
+    case TracingType::DEFAULT: {
+        auto parent = _execContext.load()->getParent();
+        for (; parent && (!parent->getBasicPlan() || !parent->getBasicPlan()->getTraceContext().has_value()); parent = parent->getParent());
```

Comment:
How is it possible the parent isn't a basic plan?

Reply:
This PR is essentially supposed to be cherry pick from https://github.com/rapyuta-robotics/alica/pull/180 so I just copied that code here. But I guess getBasicPlan() can never return nullptr so I'll get rid of it